File: course1/paradigms/Clojure/examples/magic.clj
(defn lecture [name]
  (println)
  (println "========================================")
  (println name)
  (println "========================================"))

(defn chapter [name]
  (println)
  (println "==========" name "=========="))

(defn section [name]
  (println)
  (println "----------" name "----------"))


(defn example' [prefix expression]
  (let [result (try
                 (eval expression)
                 (catch Throwable e (str (.getSimpleName (type e)) ": " (.getMessage e))))]
    (println (str "    " prefix expression " -> " result))))
(defn example
  ([description expression] (example' (str description ": ") expression))
  ([expression] (example' "" expression)))

File: course1/paradigms/javascript/examples/magic.js
"use strict";

// Magic helper functions
function example(s, description) {
    var result;
    try {
        result = eval(s);
    } catch (e) {
        result = e;
    }
    if (description) {
        println(description + ":", s, "->", result);
    } else {
        println(s, "->", result);
    }
}

function examples(collection, template) {
    collection.forEach(function(name) { 
        return example(template.replace('#', name).replace('#', name));
    });
}

function section(name) {
    println();
    println("---", name, "---");
}

function chapter(name) {
    println();
    println("==========", name, "==========");
}

function lecture(name) {
    println();
    println("===================================");
    println("=== " + name);
    println("===================================");
}

// Helper function
function dumpObject(name, o) {
    println(name);
    for (var attribute in o) {
        if (typeof(o[attribute]) == "function") {
            println("    " + attribute + "() = " + o[attribute]());
        } else {
            println("    " + attribute + " = " + o[attribute]);
        }
    }
}

